Making Detector Configuration a string and adding TestStand(TS) option

Eikenbcr opened the pull request:
- Made geometry/detector_configuration a string option that allows for HD,VD,TS detectors (others could be added easily enough)
- Added the TS dimensions to DetectorConstruction.cpp
- Made bool->String where needed

delofson0211 committed:
Merge pull request #19 from Q-Pix/develop

New Major Release (v3.0.0)

Eikenbcr committed:
Merge pull request #28 from Q-Pix/develop

Merging devlop changes to master branch

Eikenbcr committed:
Changes to DetectorConstruction to allow for TestStand (TS) geometry

Eikenbcr committed:
Merge branch 'Q-Pix:master' into feature/teststand_geometry

Eikenbcr committed:
Change detectorConfiguration to string var

Eikenbcr committed:
Merge branch 'feature/teststand_geometry' of https://github.com/Eikenbcr/qpixg4 into feature/teststand_geometry

Eikenbcr committed:
updating sample macros for detector configuration option

Eikenbcr committed:
updated GetDetectorConfiguration print statements

Eikenbcr committed:
updated GetDetectorConfiguration print statements

Eikenbcr committed:
Delete .vscode directory

Eikenbcr committed:
Fixed the dimensions of teststand

Eikenbcr commented:
I've already tested this with the various options on a single muon sample,

"/geometry/detector_configuration HD" correctly used the HD config
"/geometry/detector_configuration VD" correctly used the VD config
"/geometry/detector_configuration TS" correctly used the TS config

"/geometry/detector_configuration whatever" gave an error and didn't run the simulation:
parameter value (whatever) is not listed in the candidate List.
***** Illegal parameter (0) </geometry/detector_configuration whatever> *****

Not using the parameter correctly defaulted to the HD configuration.

kelseymh commented on src/ConfigManager.h:
```diff
@@ -69,7 +69,7 @@ class ConfigManager {
     static G4bool        GetSNTimingOn()                    { return Instance()->snTimingOn_; }
     static G4String      GetTh2Name()                       { return Instance()->th2Name_; }
-    static G4bool        GetUseHDDetectorConfiguration()    { return Instance()->useHDDetectorConfiguration_; }
+    static G4String      GetDetectorConfiguration()         { return Instance()->detectorConfiguration_; }
```
Always return non built-in data types by const reference, unless you have a *really* good reason to have to return by value.  Use `const G4String&`.

Eikenbcr replied:
Will revise this line. Thank you Mike!

kelseymh commented on src/ConfigManager.h:
```diff
@@ -106,7 +106,7 @@ class ConfigManager {
     static void SetSNTimingOn(G4bool value)                 { Instance()->snTimingOn_ = value; }
     static void SetTh2Name(G4String value)                  { Instance()->th2Name_ = value; }            
-    static void SetUseHDDetectorConfiguration(G4bool value) { Instance()->useHDDetectorConfiguration_ = value; }
+    static void SetDetectorConfiguration(G4String value)    { Instance()->detectorConfiguration_ = value; }
```
For non-built-in data types, pass by reference, never by value (unless you have a really good reason to need by-value passing).  Here, use `const G4String& value`.

Eikenbcr replied:
Will revise this line. Thank you Mike!

kelseymh commented on src/DetectorConstruction.cpp:
```diff
@@ -63,7 +73,15 @@ G4VPhysicalVolume* DetectorConstruction::Construct()
                       
 
-  std::cout << " Detector configuration is: " << ConfigManager::GetUseHDDetectorConfiguration() << std::endl;
+  if (ConfigManager::GetDetectorConfiguration() == "HD") {
+      std::cout << " Detector configuration is: HD" << std::endl;
+  } else if (ConfigManager::GetDetectorConfiguration() == "VD") {
+      std::cout << " Detector configuration is: VD" << std::endl;
+  } else if (ConfigManager::GetDetectorConfiguration() == "TS") {
+      std::cout << " Detector configuration is: TS" << std::endl;
+  } else {
+      std::cout << " Detector configuration is: DEFAULT (HD)" << std::endl;
+  }
```
Why not just one line?
```
std::cout << " Detector configuration is: " << ConfigManager::GetDetectorConfiguration() << std::endl;
```

Eikenbcr replied:
That would certainly be easier. Will change this.

kelseymh reviewed (changes requested):
Looks good!   If it was me, I'd still replace the explicit code in the if-cascade (and the instantiation of solid and LV) with member functions for each detector type.  That would allow for more complicated kinds of detector structures in the future.

Eikenbcr committed:
Making Detector Geometry options member functions

Eikenbcr committed:
moved return to world function

Eikenbcr committed:
Made WorldBox the same for all detectors

Eikenbcr committed:
added config variable

Eikenbcr committed:
removing detector function

Eikenbcr committed:
remove redundant print statement

Eikenbcr commented:
@kelseymh I fixed the issues you pointed out with the code. I also updated the detector geometries to void functions. I had some problems making the world and detector materials and positions as member functions. I wasn't able to figure out how to make their outputs global so I can use them in G4VPhysicalVolume* DetectorConstruction::Construct().

kelseymh commented:
> @kelseymh I fixed the issues you pointed out with the code. I also updated the detector geometries to void functions.

Oh, I was thinking of something slightly different; maybe we should talk in Skype and I can show you some of the CDMS code.  If you put each detector type into a "`G4LogicalVolume* BuildHDDetector() const {...}`" function, then the main `Construct()` can call that function, stash the return value into a local variable, and then in the main `Construct()` do the placement into the world.

> I had some problems making the world and detector materials and positions as member functions.

The detector material shouldn't necessarily need to be a member "function."  At best, you make a _data member_ of the class, like `G4String detectorMaterial`, and then the individual BuildXYZDetector() functions can reference it to instantiate the right `G4Material*`.  You could even (if you're overly ambitious!) put the active material into ConfigManager and let users change from argon to something else if they wanted to.  That's probably asking for trouble, though.

There's no reason that the detector builder functions need access to the world parameters.  That's the responsibility of `Construct()`.

Eikenbcr committed:
revert back to original method for detector config

Eikenbcr committed:
simplify print statement for config

Eikenbcr commented:
Okay, lets meet tomorrow to talk about it more. I've reverted back to what I originally had with the fixed print statement because I don't fully trust/understand my void function implementation.

Eikenbcr committed:
Changing detector geometries to member functions

Eikenbcr committed:
Changing detector geometries to member functions

Eikenbcr committed:
revised order of geometry in Construct() to properly get (length,width,height) of detector

Eikenbcr committed:
revert to detector.logical name

Eikenbcr committed:
revert to detector.logical name

Eikenbcr committed:
ConstructSDandField() changes for when detector_logic_vol is null

Eikenbcr committed:
Indenting lines in ConstructSDandField()

Eikenbcr committed:
Merge pull request #33 from Eikenbcr/feature/teststand_geometry

Making detector geometry member functions in DetectorConstruction.cpp

Eikenbcr commented:
I've pulled some changes that changed detector geometries into functions.
